runtime/wasm/src/host.rs

```diff
@@ -684,49 +689,122 @@ impl RuntimeHostTrait for RuntimeHost {
+                                event_handler.event,
+                                abi_name,
+                                data_source_name,
+                            ))
+                        }
+                    };
+
+                    acc.push((event_handler, event_abi));
+                    Ok(acc)
+                });
+
+        // If there are no valid handlers, fail processing the event
+        let valid_handlers = match valid_handlers {
+            Ok(valid_handlers) => valid_handlers,
             Err(e) => {
-                return Box::new(future::err(format_err!(
-                    "Failed to parse parameters of event: {}: {}",
-                    event_handler.event,
-                    e
-                )));
+                return Box::new(future::err(e)) as Box<dyn Future<Item = _, Error = _> + Send>
             }
         };
 
+        // Filter out handlers whose corresponding event ABIs cannot decode the
+        // params (this is common for overloaded events that have the same topic0
+        // but have indexed vs. non-indexed params that are encoded differently).
+        //
+        // Map (handler, event ABI) pairs to (handler, decoded params) pairs.
+        let mut matching_handlers = valid_handlers
+            .into_iter()
+            .filter_map(|(event_handler, event_abi)| {
+                event_abi
+                    .parse_log(RawLog {
+                        topics: log.topics.clone(),
+                        data: log.data.clone().0,
+                    })
+                    .map(|log| log.params)
+                    .map_err(|e| {
+                        info!(
+                            logger,
+                            "Skipping handler because the event parameters do not \
+                             match the event signature. This is typically the case \
+                             when parameters are indexed in the event but not in the \
+                             signature or the other way around";
+                            "handler" => &event_handler.handler,
+                            "event" => &event_handler.event,
+                            "error" => format!("{}", e),
+                        );
+                    })
+                    .ok()
+                    .map(|params| (event_handler, params))
+            })
+            .collect::<Vec<_>>();
+
+        // If none of the handlers match the event params, fail processing the event
+        if matching_handlers.is_empty() {
+            return Box::new(future::err(format_err!(
+                "No matching event handlers found for in data source \"{}\"",
+                data_source_name,
+            )));
+        }
+
+        // Process the event with the matching handler
+        let (event_handler, params) = matching_handlers.pop().unwrap();
+
+        // Fail if there is more than one matching handler
+        if matching_handlers.len() > 1 {
```

Comment:
`pop` will remove the element, so the correct check here is `!matching_handlers.is_empty()`.

Reply:
Oops, yes, indeed! I moved the `pop` up there after I wrote the length check.